src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -96,126 +369,135 @@ pub fn copy_to<G, F>(
+                            if !up_to.less_equal(&ts) {
+                                if diff < 0 {
+                                    Err(anyhow::anyhow!(
+                                        "Invalid data in source errors, saw retractions ({}) for \
+                                        row that does not exist",
+                                        diff * -1,
+                                    ))?;
+                                }
+                                row_count += u64::try_from(diff).unwrap();
+                                let uploader = s3_uploaders.entry(batch).or_insert_with(|| {
+                                    debug!(%sink_id, %worker_id, "handling batch: {}", batch);
+                                    CopyToS3Uploader::new(
+                                        sdk_config.clone(),
+                                        connection_details.clone(),
+                                        &sink_id,
+                                        batch,
+                                    )
                                 });
-                            for _ in 0..diff {
-                                match uploader.append_row(&row).await {
-                                    Ok(()) => {}
-                                    Err(e) => {
-                                        onetime_callback(Err(e.to_string()));
-                                        return;
-                                    }
+                                for _ in 0..diff {
+                                    uploader.append_row(&row).await?;
                                 }
                             }
                         }
                     }
-                }
-                AsyncEvent::Progress(frontier) => {
-                    if PartialOrder::less_equal(&up_to, &frontier) {
-                        for uploader in s3_uploaders.values_mut() {
-                            match uploader.flush().await {
-                                Ok(()) => {}
-                                Err(e) => {
-                                    onetime_callback(Err(e.to_string()));
-                                    return;
-                                }
+                    AsyncEvent::Progress(frontier) => {
+                        if PartialOrder::less_equal(&up_to, &frontier) {
+                            for uploader in s3_uploaders.values_mut() {
+                                uploader.flush().await?;
                             }
+                            // We are done, send the final count.
+                            return Ok(row_count);
                         }
-                        // We are done, send the final count.
-                        onetime_callback(Ok(row_count));
-                        return;
                     }
                 }
             }
-        }
+            Ok::<u64, anyhow::Error>(row_count)
+        };
```

Comment:
If you make this a bare async block you can `.await` it here
```suggestion
        }.await;
```